pkg/election/lease.go

```diff
@@ -165,11 +213,22 @@ func (l *lease) keepAliveWorker(ctx context.Context, interval time.Duration) <-c
 				if l.ID.Load() != nil {
 					leaseID = l.ID.Load().(clientv3.LeaseID)
 				}
+				l.addUnHealthyTimesLock(start)
+				// add failpoint to simulate etcd is unhealthy.
+				failpoint.Inject("timeoutKeepAliveOnce", func() {
+					log.Info("timeoutKeepAliveOnce is injected, sleep 1s to simulate etcd is unhealthy")
+					time.Sleep(unhealthyTimesRecordTimeout)
+				})
+
 				res, err := l.lease.KeepAliveOnce(ctx1, leaseID)
 				if err != nil {
 					log.Warn("lease keep alive failed", zap.String("purpose", l.Purpose), zap.Time("start", start), errs.ZapError(err))
 					return
 				}
+
+				if time.Since(start) < unhealthyTimesRecordTimeout {
+					l.removeUnHealthyTimesLock(start)
```

Comment:
Could we only add the unhealthy time when it's greater than `unhealthyTimesRecordTimeout` rather than adding it first and then deleting it?

Reply:
Yes, we can. But it will rely on upper layer timeout mechanism which reaches 3s(or greater)

add and then remove will be more fine-grained.

But I am not sure keeping fine-grained under 3s is necessary